keps/sig-auth/20190926-dynamic-audit-policy.md

```diff
@@ -0,0 +1,387 @@
+
+Requests matched by rule are *selected*. Rules define the criteria to match requests. The way it is interpreted is no different to the file policies.
+
+Selection rules are typed:
+- `requestSelectors` general rules selecting requests by user and/or verb for example
+- `nonResourceSelectors` rules selecting non-resource requests
+- `resourceSelectors` rules selecting resource requests
+- `clusterResourceSelectors` rules selecting cluster resource requests
+
+In contrast to the existing `audit.k8s.io/v1` `Policy` where non-resource request selectors are specified with the same rule model as resource requests selectors, in `AuditClass` objects they are no longer mixed. Having a dedicated selector model makes `AuditClass` objects easier to comprehend and less error-prone, by sparing them from implicit rules such as invalid coexistence of `nonResourceUrl` and `namespace`/`resource` elements.
+
+### Dynamic Audit Policy Runtime
+
+#### Getting Ready for Audit
+
+When an `AuditSink` configuration is registered, its policy is compiled into a list of rules composed of the rules from the referenced `AuditClass` objects with their effective levels assigned. The compiled policy is then applied to the dynamic backend, maintained by the API server audit plugin for the corresponding `AuditSink`. At runtime, the `AuditSink` dynamically adapts to changes in the `AuditSink` policy or the `AuditClass` objects that it references, by recompiling and applying the new effective policy.
+
+> The compiled effective policy is an internal model and the `audit.k8s.io/v1` `Policy` was chosen to play that role. More details on how that dynamic audit machinery plugs in the server auditing are available in the Dynamic Control KEP's [Policy Enforcement](https://github.com/kubernetes/enhancements/blob/f1a799d5f4658ed29797c1fb9ceb7a4d0f538e93/keps/sig-auth/0014-dynamic-audit-configuration.md#policy-enforcement) section. 
+
+#### Runtime Control on References
+
+Introducing a dependency in a `AuditSink` policy to another runtime object (`AuditClass`) by reference, raises concerns about maintaining the consistency of this relationship. For example, what should be the status of an `AuditSink` comprising a policy that refers to a non-existing `AuditClass`. How is this addressed when it is created and when the `AuditClass` is deleted at runtime? 
+
+This proposal borrows conceptually from the approach of `Pod` resources towards used resources, such as secrets. `AuditSink` objects are created even if referenced `AuditClass` objects are not yet available at the time, but the corresponding dynamic backend does not send events to that sink until all references are available. Deleting an `AuditClass` object without removing existing policy references to it is blocked.
```

Comment:
That's probably where we have to start. Let's elaborate.   

> Audit classifier labels the request with all audit classes that it **matches**

What's the criteria for a classifier to match a request to a class dynamically at runtime? A request has no supplementary metadata. So the only thing left is to figure out if a class has a rule that matches the request. Then again, how's that different to step 3 and what is already proposed here?

Considering the context, I would understand the proposal if on step 2 it was "Audit classifier labels the request with all audit classes". Was that what you actually meant by chance?

Reply:
I'm confused abut the _labels_ part too. So you are saying that class (rule sets) references in a policy should not be normative, but kind of recommendation? If classes are delivered by one role (say an _operator_, together with the deployment of application or extension bundles), and the policies are defined by a more privileged role (say an _administrator_), that means that operators can influence what's being audited. A proper delete of a class could hide their activities for good. And that would not sit well if you wanted only administrators to control on auditing.

Reply:
Let me try to give an example to illustrate what I was proposing:

Suppose I have the following audit classes (pseudo-code):

```yaml
- name: tim-requests
  users: ["tim"]
- name: bob-requests
  users: ["bob"]
- name: get-requests
  verbs: ["get"]
- name: secrets-requests
  resources: ["secrets"]
- name: pod-requests
  resources: ["pods"]
```

And my audit policy looks like:
```yaml
- level: None
  class: "foo-requests" // we don't have "foo-requests"
- level: Metadata
  class: "secrets-requests"
- level: RequestResponse
  class: "tim-requests"
- level: None
  class: "get-requests"
```

**1. Request Received:** Then a request comes in:
```
[user=tim] GET /api/v1/namespaces/default/secrets/foo
```

**2. Classification:** The system "labels" (attaches to the context) the matching audit classes:
```
- tim-requests
- get-requests
- secrets-requests
```

**3. Policy Decision:** Then the audit policy takes the first matching rule: `secrets-requests --> Metadata`

The classification step (2) has no notion of a "foo-requests" class, so it doesn't care about it. It only considers the classes that are defined. The policy decision step (3) doesn't care what audit classes are defined, only what classes the request is labeled with. The request isn't labeled with "foo-requests" just like it isn't labeled with "bob-requests" - from the policy decision perspective there's no difference between those 2 situations.

Does this help clarify?

Reply:
Yes, thank you. That is actually inline with the [proto-implementation ](https://github.com/kubernetes/kubernetes/commit/3815ae4e9486cbfaa3ec4c69caef266e158de370) that was prepared for the initial policy proposal (except for subtleties such as classes attached to context but at the end it works as described) an I can imagine a good deal of it will be reused. Let me confirm a few things with you though:
- Step 2 does not occur on per-request basis. I am using a cached and synchronized AuditClasses state (using informers) to figure out whats available. I don't check the state synchronously with `list` per request. That would be an overkill.
- When I refer to an _effective policy_ or _compiled policy_ in the KEP I mean a policy that is produced out of the classes identified on step 2 and the policy levels, and then normalized in a ready state for step 3. It is updated on changes in the audit sink or (referenced) classes.
- For step 3 I use the legacy checker mechanism in server auditing to figure out level and stage out of the compiled policy and the request.

It's seems we should focus the discussion on who controls what (my second comment). I am not completely fond of the options for resources and responsibilities breakdown as they stand now. 

I'm starting to think we should elaborate the following. An admin can strictly define the cluster resources they do care about and operators can simultaneously contribute domain-specific rules together with audit levels that are then dynamically composed in a policy, should an admin allow that. Consider my proposal for composition by labels/annotations as an enabler for that. Of course we need to decide on the ordering and levels issues then. But I think that nicely solves the responsibilities mix problem that we have now. WDYT?

Reply:
Sorry, I'm having a hard time following this thread. Let me try to recap the issue:

1. Administrators should be able to define an audit policy in a secure way that can only be altered by other admins or highly privileged users (e.g. security team).
2. Operators want to be able to extend the audit policy for a service they operate. For example, I added a new type and I want to make sure it is audited appropriately, or I added a new service that makes a lot of requests, and I want to omit it to cut down on noise.

I think the problem is that these 2 use-cases are fundamentally at odds. I suppose there might be some way for the admin to define a policy on how/what the operator can modify, but I can't think of a good way to define such a policy. Ultimately, I think something like gatekeeper or OPA might be the best tool here, to allow arbitrary restrictions on what audit classes a user can define. IMO we should focus on the first case.

There is also a third use case:
3) I want to add some special-case logging for my service, e.g. for debugging.

As it stands, there isn't a good way for an administrator to allow this securely. Even if a user is allowed to edit a webhook created for them, they could abuse that to read everything (secrets) in the cluster.  In this case, I think what we really need is namespaced audit webhooks, and we should consider that out-of-scope for this KEP.

Does this answer your questions, or am I way off base here?

Reply:
You are not way off, but we are still thinking in different terms how to achieve it. I am convinced that we can address the scenarios with clear separation and simpler means, with the policy "ease of reasoning" being the only open topic. I think I've laid out my point sufficiently in my last post so I will not repeat it. I suggest we make a touch base call to find the common ground as doing that offline seems to go nowhere.